cmd/libs/go2idl/make-client/generators/client-generator.go

```diff
@@ -0,0 +1,270 @@
+	}
+}
+
+func (g *genClient) Imports(c *generator.Context) (imports []string) {
+	return append(g.imports.ImportLines())
+}
+
+// GenerateType makes the body of a file implementing a set for type t.
+func (g *genClient) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+	pkg := filepath.Base(t.Name.Package)
+	m := map[interface{}]interface{}{
+		"type":    t,
+		"package": pkg,
+		"Package": namer.IC(pkg),
+		// @lavalamp: how about making a plural function in pkg/util, RESTMapper also needs such a function.
+		// m["plural"] = util.Plural(type.Name.Name)
+		"fieldSelector":    fieldSelector,
+		"labelSelector":    labelSelector,
+		"watchInterface":   watchInterface,
+		"apiDeleteOptions": apiDeleteOptions,
+		"apiListOptions":   apiListOptions,
+	}
+	sw.Do(nameSpacerTemplate, m)
+	sw.Do(interfaceTemplate, m)
+	sw.Do(structTemplate, m)
+	sw.Do(newStructTemplate, m)
+	sw.Do(listTemplate, m)
+	sw.Do(getTemplate, m)
+	sw.Do(deleteTemplate, m)
+	sw.Do(createTemplate, m)
+	sw.Do(updateTemplate, m)
+	sw.Do(updateStatusTemplate, m)
+	sw.Do(watchTemplate, m)
+
+	return sw.Error()
+}
+
+// template for nameSpacer
+var nameSpacerTemplate = `
+// $.type|public$Namespacer has methods to work with $.type|public$ resources in a namespace
+type $.type|public$Namespacer interface {
+	$.type|public$s(namespace string) $.type|public$Interface
+}
+`
+
+// template for the Interface
+var interfaceTemplate = `
+// $.type|public$Interface has methods to work with $.type|public$ resources.
+type $.type|public$Interface interface {
+	List(label $.labelSelector|raw$, field $.fieldSelector|raw$) (*$.type|raw$List, error)
+	Get(name string) (*$.type|raw$, error)
+	Delete(name string, options *$.apiDeleteOptions|raw$) error
+	Create(*$.type|raw$) (*$.type|raw$, error)
+	Update(*$.type|raw$) (*$.type|raw$, error)
+	UpdateStatus(*$.type|raw$) (*$.type|raw$, error)
+	Watch(label $.labelSelector|raw$, field $.fieldSelector|raw$, opts $.apiListOptions|raw$) ($.watchInterface|raw$, error)
+}
+`
+
+// template for the struct that implements the interface
+var structTemplate = `
+// $.type|private$s implements $.type|public$Interface
+type $.type|private$s struct {
+	client *$.Package$Client
+	ns     string
+}
+`
+var newStructTemplate = `
+// new$.type|public$s returns a $.type|public$s
+func new$.type|public$s(c *ExtensionsClient, namespace string) *$.type|private$s {
+	return &$.type|private$s{
+		client: c,
+		ns:     namespace,
+	}
+}
+`
+var listTemplate = `
+// List takes label and field selectors, and returns the list of $.type|public$s that match those selectors.
+func (c *$.type|private$s) List(label $.labelSelector|raw$, field $.fieldSelector|raw$) (result *$.type|raw$List, err error) {
+	result = &$.type|raw$List{}
+	err = c.client.Get().Namespace(c.ns).Resource("$.type|private$s").LabelsSelectorParam(label).FieldsSelectorParam(field).Do().Into(result)
```

Comment:
nit: what do you think about separating this line into few lines (similarly to Watch() function) to make it more readable?